src/group_theory/coproduct.lean

```diff
@@ -0,0 +1,235 @@
+/-
+Copyright (c) 2021 David Wärn. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: David Wärn
+-/
+import data.list.chain
+/-!
+# The coproduct of a family of groups
+
+We work with an `ι`-indexed family `G` of monoids, where both `ι` and all `G i` have decidable
+equality. We define the coproduct `coprod G` of `G` as the type of reduced words. One could define
+this more abstractly in terms of its presentation, but this definition has the advantage of giving
+decidable equality.
+
+When `G i` are all groups, we show that `coprod G` is also a group.
+
+## Main definitions
+
+- `coprod G`
+- `of {i}`: the embedding `G i →* coprod G`.
+- `ump X : (Π {i}, G i →* X) ≃ (coprod G →* X)`: the universal mapping property of the coproduct.
+
+-/
+
+open list
+
+section monoids
+
+variables {ι : Type*} (G : Π i : ι, Type*) [Π i, monoid (G i)]
+  [decidable_eq ι] [∀ i, decidable_eq (G i)]
+
+@[derive decidable_eq]
+def coprod : Type* := { l : list (Σ i, { g : G i // g ≠ 1 }) // (l.map sigma.fst).chain' (≠) }
+variable {G}
+
+-- `w.head_isn't i` says that `i` is not the head of `w`.
+def coprod.head_isn't (w : coprod G) (i : ι) : Prop := ∀ p ∈ (w.val.map sigma.fst).head', i ≠ p
```

Comment:
I'd not though of using `'` as an apostrophe in a lemma name!

Reply:
It's maybe a bit unorthodox :-). A lot of the names here are really placeholders.